### ISSUE TITLE
Zed crashes when hovering specific inline information about compiler errors when programming with Rust

### ISSUE THREAD

**Issue by Nicceboy** [zed-industries/zed#16656]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

When programming Rust on Mac with default configuration (expect vim keybindings), Zed crashes on a certain case when hovering the underlined compiler error area.

Crash happens immediately when I move the mouse pointer to the specific underlined area to see the inline popup. 

I have attached the panic log below. I also managed to minimize the crashing code, where the crash happens when hovering `Self::Variable(*lhs + rhs)` part, inside brackets.

```rust
﻿use num_bigint::BigInt;
#[derive(Debug, Clone, Ord, Hash, Eq, PartialEq, PartialOrd)]
pub enum Integer {
    Primitive(isize),
    Variable(Box<BigInt>),
}

macro_rules! impl_add_integer {
    ($($t:ty),*) => {
        $(
            impl core::ops::Add<$t> for Integer {
                type Output = Self;
                fn add(self, rhs: $t) -> Self::Output {
                    match self {
                        Self::Primitive(lhs) => {
                            let result = lhs.checked_add(rhs as isize);
                            match result {
                                Some(value) => Self::Primitive(value),
                                None => Self::Variable(Box::new(BigInt::from(lhs) + rhs)),
                            }
                        }
                        Self::Variable(lhs) => {
                            Self::Variable(*lhs + rhs)
                        }
                    }
                }
            }
        )*
    };
}
impl_add_integer!(u8);
```

### Environment

Zed: v0.149.3 (Zed)
OS: macOS 14.5.0
Memory: 32 GiB
Architecture: aarch64

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary><pre>

{
  "thread": "main",
  "payload": "byte index 384 is out of bounds of `rustc: mismatched types\nexpected struct `Box<BigInt>`\n   found struct `BigInt`\nfor more on the distinction between the stack and the heap, read https://doc.rust-lang.org/book/ch15-01-box.html, https://doc.rust-lang.org/rust-by-example/std/box.html, and htt`[...]",
  "location_data": {
    "file": "crates/markdown/src/markdown.rs",
    "line": 698
  },
  "backtrace": [
    "<backtrace::capture::Backtrace>::create",
    "<backtrace::capture::Backtrace>::new",
    "zed::reliability::init_panic_hook::{closure#0}",
    "std::panicking::rust_panic_with_hook",
    "std::panicking::begin_panic_handler::{{closure}}",
    "std::sys_common::backtrace::__rust_end_short_backtrace",
    "_rust_begin_unwind",
    "core::panicking::panic_fmt",
    "core::str::slice_error_fail_rt",
    "core::str::slice_error_fail",
    "<markdown::MarkdownElement as gpui::element::Element>::request_layout",
    "<gpui::element::Drawable<markdown::MarkdownElement>>::request_layout",
    "<gpui::element::Drawable<gpui::view::View<markdown::Markdown>>>::request_layout",
    "<gpui::elements::div::Div as gpui::element::Element>::request_layout::{closure#0}::{closure#0}",
    "<gpui::elements::div::Interactivity>::request_layout::<<gpui::elements::div::Div as gpui::element::Element>::request_layout::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::request_layout",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>>>::request_layout",
    "<gpui::elements::div::Div as gpui::element::Element>::request_layout::{closure#0}::{closure#0}",
    "<gpui::elements::div::Interactivity>::request_layout::<<gpui::elements::div::Div as gpui::element::Element>::request_layout::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::request_layout",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>>>::request_layout",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::layout_as_root",
    "<editor::element::EditorElement as gpui::element::Element>::prepaint::{closure#0}::{closure#0}::{closure#0}",
    "<editor::element::EditorElement as gpui::element::Element>::prepaint::{closure#0}",
    "<editor::element::EditorElement as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<editor::element::EditorElement> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::view::AnyView> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint::{closure#0}",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::view::AnyView> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::element::Drawable<gpui::element::AnyElement> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}::{closure#0}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}::{closure#1}",
    "<gpui::elements::div::Interactivity>::prepaint::<core::option::Option<gpui::window::Hitbox>, <gpui::elements::div::Div as gpui::element::Element>::prepaint::{closure#0}>::{closure#0}",
    "<gpui::elements::div::Div as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::elements::div::Stateful<gpui::elements::div::Div>> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::view::AnyView as gpui::element::Element>::prepaint",
    "<gpui::element::Drawable<gpui::view::AnyView> as gpui::element::ElementObject>::prepaint",
    "<gpui::element::AnyElement>::prepaint",
    "<gpui::window::WindowContext>::draw",
    "<gpui::window::Window>::new::{closure#4}::{closure#1}",
    "<gpui::window::Window>::new::{closure#4}",
    "gpui::platform::mac::window::step",
    "<gpui::platform::mac::platform::MacPlatform as gpui::platform::Platform>::run",
    "<gpui::app::App>::run::<zed::main::{closure#3}>",
    "zed::main",
    "std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>",
    "std::rt::lang_start::<()>::{closure#0}",
    "std::rt::lang_start_internal",
    "_main"
  ],
  "app_version": "0.149.3",
  "release_channel": "Zed",
  "os_name": "macOS",
  "os_version": "14.5.0",
  "architecture": "aarch64",
  "panicked_on": 1724314871741,
  "installation_id": "775f5793-47c7-4a04-b616-a2afe2e95cc2",
  "session_id": "44a134f5-7846-4fc6-9d4a-d324368cfc64"
}
</pre></details>

**Comment by SomeoneToIgnore** [comment]
I believe that's fixed in https://github.com/zed-industries/zed/pull/16601 which was not released just yet.

**Comment by Nicceboy** [comment]
> I believe that's fixed in #16601 which was not released just yet.

Yep, I compiled the main branch and that indeed seems to be case.
